pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConnectionHandler.java

```diff
@@ -63,11 +66,20 @@ protected void grabCnx() {
             return;
         }
 
-        if (!isValidStateForReconnection()) {
-            // Ignore connection closed when we are shutting down
-            log.info("[{}] [{}] Ignoring reconnection request (state: {})",
-                    state.topic, state.getHandlerName(), state.getState());
-            return;
+        synchronized (this) {
+            if (duringConnect) {
+                log.info("[{}] [{}] Skip grabbing the connection since there is a pending connection",
+                        state.topic, state.getHandlerName());
+                return;
+            }
+            if (isValidStateForReconnection()) {
```

Comment:
Why? It will be skipped when checking `duringConnect`

Reply:
When the second task gets the lock, the first task maybe is finished.

Reply:
Okay, I use an atomic boolean instead. PTAL gain.

Reply:
The purpose of this PR is to avoid more than one `grabCnx` running at the same time, right? And does not address the execution of `grabCnx` even if the state is `Ready`, right?

Reply:
@BewareMyPower  and this comment

Reply:
> And does not address the execution of grabCnx even if the state is Ready

Yes. What does it affect?

Reply:
After the state is changed to `Ready`, the consumer will receive messages. The next `grabCnx` will make the received messages lost

Reply:
> After the state is changed to Ready, the consumer will receive messages. The next grabCnx will make the received messages lost

This seems another issue existing in the current codebase instead of introduced by this patch?

Reply:
> This seems another issue existing in the current codebase instead of introduced by this patch?

After talking with @BewareMyPower , another pr will fix the problem. I will reopen https://github.com/apache/pulsar/pull/20591 to fix it.